axon-cosmosdb/src/main/java/org/axonframework/extensions/cosmosdb/eventsourcing/tokenstore/CosmosTokenStore.java

```diff
@@ -0,0 +1,516 @@
+            throw new UnableToInitializeTokenException("Could not initialize token.", e);
+        }
+    }
+
+    @Override
+    public void deleteToken(@Nonnull String processorName, int segment) throws UnableToClaimTokenException {
+        try {
+            getCurrent(processorName, segment)
+                    .flatMap(current -> {
+                        String owner = current.getItem().getOwner();
+                        if (!nodeId.equals(owner)) {
+                            return Mono.error(new UnableToClaimTokenException(
+                                    String.format(CANT_DELETE_MESSAGE_FORMAT, owner, nodeId)));
+                        }
+                        return getContainer(processorName).deleteAllItemsByPartitionKey(
+                                new PartitionKey(Integer.toString(segment)),
+                                new CosmosItemRequestOptions().setIfMatchETag(null)
+                        );
+                    })
+                    .block();
+        } catch (UnableToClaimTokenException e) {
+            throw e;
+        } catch (Exception e) {
+            throw new UnableToClaimTokenException("Could not delete token.", e);
+        }
+    }
+
+    @Override
+    public boolean requiresExplicitSegmentInitialization() {
+        return true;
+    }
+
+    @Override
+    public int[] fetchSegments(@Nonnull String processorName) {
+        List<Integer> segments = getContainer(processorName)
+                .queryItems(ALL_QUERY, CosmosTokenItem.class)
+                .map(CosmosTokenItem::getSegment)
+                .collectList()
+                .block();
+        if (isNull(segments)) {
+            return new int[0];
+        }
+        // toArray doesn't work because of autoboxing limitations
+        int[] ints = new int[segments.size()];
+        for (int i = 0; i < ints.length; i++) {
+            ints[i] = segments.get(i);
+        }
+        return ints;
+    }
+
+    @Override
+    public List<Segment> fetchAvailableSegments(@Nonnull String processorName) {
+        int[] all = fetchSegments(processorName);
+        return getContainer(processorName)
+                .queryItems(ALL_QUERY, CosmosTokenItem.class)
+                .filter(this::claimable)
+                .map(item -> Segment.computeSegment(item.getSegment(), all))
+                .collectList()
+                .block();
+    }
+
+    private CosmosAsyncDatabase createDatabaseIfNotExists(String databaseName) {
+        return client
+                .createDatabaseIfNotExists(databaseName)
+                .map(r -> client.getDatabase(r.getProperties().getId()))
+                .block();
+    }
+
+    private CosmosAsyncContainer createContainerIfNotExists(String processorName) {
+        return database.createContainerIfNotExists(processorName, ID_PARTITION_KEY)
+                       .map(r -> database.getContainer(r.getProperties().getId()))
+                       .block();
+    }
```

Comment:
Ah, have you done this to ease the `ALL_QUERY` perhaps? If so, I would guess the query could also validate the used `processorName`, right?

Reply:
Mostly to keep things simple and, I guess/hope, more performant. Yes, we could also make it so that it would use a single container, add the processor name to the `CosmosTokenItem`, add the processor name to the id, and adopt the query. I do think it makes sense to make the default database name 'axon-tokenstore' instead. WDYT?

Reply:
Going to change this, as there can be a limit on the number of containers to use.